sklearn/tree/tree.py

```diff
@@ -64,6 +66,119 @@
 SPARSE_SPLITTERS = {"best": _splitter.BestSparseSplitter,
                     "random": _splitter.RandomSparseSplitter}
 
+
+# =============================================================================
+# Support functions
+# =============================================================================
+def preproc_categorical(X, categorical, check_input):
```

Comment:
Well, I can't make these functions private because they are used by ensembles (forest, gradient boosting).

`LabelEncoder` doesn't have quite the desired behavior. If I call `transform()` with category labels that were not present when fitting the encoder, it raises a `ValueError`. What I really want is to accept unfamiliar category labels and just encode them to higher values than already exist in its transform. (This is so that people can predict using samples containing categories not trained on.)

I'm not super happy with this code anyway, and I'm considering just forcing users to preprocess their categorical feature values into small integers outside of `DecisionTree`. After all, it's not the job of decision trees to preprocess the data. Removing this functionality would greatly simplify the changes in this PR and would reduce surface area for bugs.